File: examples/python/webhook_receiver.py
"""
Veri Merkezi Webhook Receiver — Python / Flask

Setup:
    pip install flask
    export VM_WEBHOOK_SECRET=whsec_xxx
    python webhook_receiver.py
"""
import os
import hmac
import json
import time
import hashlib
import logging
from flask import Flask, request, abort, jsonify

app = Flask(__name__)
logging.basicConfig(level=logging.INFO, format='%(asctime)s [%(levelname)s] %(message)s')

SECRET = os.getenv('VM_WEBHOOK_SECRET', 'whsec_REPLACE_ME').encode()
processed_events = set()  # Production'da Redis/DB


@app.route('/wa-webhook', methods=['POST'])
def webhook():
    signature = request.headers.get('X-VeriMerkezi-Signature-256', '')
    timestamp = request.headers.get('X-VeriMerkezi-Timestamp', '')
    event_id = request.headers.get('X-VeriMerkezi-Event-Id', '')
    raw_body = request.get_data(as_text=True)

    # 1) Replay koruması
    try:
        if abs(time.time() - int(timestamp)) > 300:
            app.logger.warning(f'expired timestamp: {timestamp}')
            abort(403, 'expired')
    except ValueError:
        abort(403, 'invalid_timestamp')

    # 2) İmza doğrulama
    expected = 'sha256=' + hmac.new(
        SECRET,
        f'{timestamp}.{raw_body}'.encode(),
        hashlib.sha256
    ).hexdigest()

    if not hmac.compare_digest(expected, signature):
        app.logger.warning(f'invalid signature for event_id={event_id}')
        abort(403, 'invalid_signature')

    # 3) Idempotency
    if event_id in processed_events:
        return jsonify({'ok': True, 'note': 'already_processed'})

    # 4) Parse + dispatch
    try:
        event = json.loads(raw_body)
    except json.JSONDecodeError:
        abort(400, 'invalid_json')

    try:
        event_type = event.get('event', '')
        data = event.get('data', {})

        if event_type == 'message.received':
            handle_incoming_message(data)
        elif event_type == 'message.echo':
            # CoExistence: işletme telefondaki WhatsApp uygulamasından müşteriye yazdı
            app.logger.info(f'İşletme yanıtladı -> {data.get("to")}: {data.get("text") or ""}')
        elif event_type in ('message.status.sent', 'message.status.delivered', 'message.status.read'):
            handle_status_update(data, event_type)
        elif event_type == 'message.status.failed':
            handle_failed_message(data)
        elif event_type == 'template.approved':
            app.logger.info(f'Şablon onaylandı: {data.get("template_name")}')
        elif event_type in ('template.rejected', 'template.flagged', 'template.paused'):
            durum = {'template.rejected': 'reddedildi', 'template.flagged': 'işaretlendi', 'template.paused': 'durduruldu'}[event_type]
            sebep = f' · {data["reason"]}' if data.get('reason') else ''
            app.logger.warning(f'Şablon {durum}: {data.get("template_name")} ({data.get("language")}){sebep}')
        elif event_type == 'quality.changed':
            app.logger.info(f'Kalite: {data.get("phone")} -> {data.get("quality")}')
        elif event_type == 'account.alert':
            # field: account_update | account_alerts — event: Meta'nın olay adı (örn. DISABLED_UPDATE)
            app.logger.warning(f'Hesap uyarısı: {data.get("field")} · {data.get("event")}')
        else:
            app.logger.info(f'Bilinmeyen event: {event_type}')

        processed_events.add(event_id)
        return jsonify({'ok': True})

    except Exception as e:
        app.logger.error(f'handler error: {e}')
        return jsonify({'ok': False, 'error': 'handler_error'}), 500


# ═══════════════════════════════════════════════════════════════
# HANDLER'LAR — Kendi iş mantığınızı buraya yazın
# ═══════════════════════════════════════════════════════════════

def handle_incoming_message(data: dict) -> None:
    # data['text'] düz metindir (medyada açıklama; açıklama yoksa boş).
    # Telefonu gizli kullanıcılarda data['from'] None gelir; kimlik data['user_id'] (BSUID) olur.
    from_ = data.get('from') or data.get('user_id') or ''
    text = data.get('text') or ''
    msg_type = data.get('type') or 'text'
    contact_name = data.get('name') or (data.get('contact') or {}).get('profile_name') or 'Müşteri'
    media = data.get('media') or {}

    if media.get('media_id'):
        app.logger.info(f'{contact_name} ({from_}) {msg_type} gönderdi · media_id={media["media_id"]} {text}'.rstrip())
        # Dosyayı indirmek için (bkz. docs/10-media.md):
        # vm = VeriMerkeziClient(os.getenv('VM_API_KEY'))
        # vm.download_media(media['media_id'], f'medya-{media["media_id"]}')
    elif media.get('error'):
        # Nadiren dosya Meta'dan alınamaz; bu durumda indirilebilir dosya yoktur.
        app.logger.warning(f'{contact_name} ({from_}) {msg_type} gönderdi, dosya alınamadı: {media["error"]}')
    else:
        app.logger.info(f'{contact_name} ({from_}): {text}')

    # Otomatik yanıt göndermek için:
    # from verimerkezi import VeriMerkeziClient
    # vm = VeriMerkeziClient(os.getenv('VM_API_KEY'))
    # vm.send_text(data['phone_number_id'], from_, 'Mesajınızı aldık!')


def handle_status_update(data: dict, event: str) -> None:
    status = event.replace('message.status.', '')
    app.logger.info(f' {data.get("wamid")} -> {status}')


def handle_failed_message(data: dict) -> None:
    # errors: Meta'nın hata listesi — [{code, title, message, error_data: {details}, href}]
    hata = (data.get('errors') or [{}])[0]
    mesaj = hata.get('message') or hata.get('title') or 'bilinmiyor'
    app.logger.error(f'HATA: {data.get("wamid")} · alıcı={data.get("recipient")} · kod={hata.get("code")} · {mesaj}')


@app.route('/health')
def health():
    return jsonify({'status': 'ok'})


if __name__ == '__main__':
    port = int(os.getenv('PORT', 3000))
    app.logger.info(f'Veri Merkezi webhook receiver: http://localhost:{port}/wa-webhook')
    app.run(host='0.0.0.0', port=port, debug=False)

File: examples/nodejs/webhook-receiver.js
/**
 * Veri Merkezi Webhook Receiver — Node.js / Express
 *
 * Setup:
 * npm install express
 * VM_WEBHOOK_SECRET=whsec_xxx node webhook-receiver.js
 */

const express = require('express');
const crypto = require('crypto');

const app = express();
const PORT = process.env.PORT || 3000;
const SECRET = process.env.VM_WEBHOOK_SECRET || 'whsec_REPLACE_ME';

// İmza doğrulaması için raw body lazım
app.use('/wa-webhook', express.raw({ type: 'application/json', limit: '5mb' }));

const processedEvents = new Set(); // Production'da Redis/DB

app.post('/wa-webhook', (req, res) => {
 const signature = req.headers['x-verimerkezi-signature-256'] || '';
 const timestamp = req.headers['x-verimerkezi-timestamp'] || '';
 const eventId = req.headers['x-verimerkezi-event-id'] || '';
 const rawBody = req.body.toString('utf8');

 // 1) Replay koruması
 if (Math.abs(Date.now() / 1000 - parseInt(timestamp, 10)) > 300) {
 console.warn('[WaWebhook] expired:', timestamp);
 return res.status(403).send('expired');
 }

 // 2) İmza doğrulama
 const expected = 'sha256=' + crypto
 .createHmac('sha256', SECRET)
 .update(`${timestamp}.${rawBody}`)
 .digest('hex');

 const sigBuf = Buffer.from(signature);
 const expBuf = Buffer.from(expected);
 if (sigBuf.length !== expBuf.length || !crypto.timingSafeEqual(sigBuf, expBuf)) {
 console.warn('[WaWebhook] invalid signature for event:', eventId);
 return res.status(403).send('invalid_signature');
 }

 // 3) Idempotency
 if (processedEvents.has(eventId)) {
 return res.json({ ok: true, note: 'already_processed' });
 }

 // 4) Parse + dispatch
 let event;
 try {
 event = JSON.parse(rawBody);
 } catch (e) {
 return res.status(400).send('invalid_json');
 }

 try {
 switch (event.event) {
 case 'message.received':
 handleIncomingMessage(event.data);
 break;
 case 'message.echo':
 // CoExistence: işletme telefondaki WhatsApp uygulamasından müşteriye yazdı
 console.log(`İşletme yanıtladı -> ${event.data.to}: ${event.data.text || '[' + event.data.type + ']'}`);
 break;
 case 'message.status.sent':
 case 'message.status.delivered':
 case 'message.status.read':
 handleStatusUpdate(event.data, event.event);
 break;
 case 'message.status.failed':
 handleFailedMessage(event.data);
 break;
 case 'template.approved':
 console.log('Şablon onaylandı:', event.data.template_name);
 break;
 case 'template.rejected':
 case 'template.flagged':
 case 'template.paused': {
 const durum = { 'template.rejected': 'reddedildi', 'template.flagged': 'işaretlendi', 'template.paused': 'durduruldu' }[event.event];
 console.log(`Şablon ${durum}: ${event.data.template_name} (${event.data.language})${event.data.reason ? ' · ' + event.data.reason : ''}`);
 break;
 }
 case 'quality.changed':
 console.log(`Kalite: ${event.data.phone} -> ${event.data.quality}`);
 break;
 case 'account.alert':
 // field: account_update | account_alerts — event: Meta'nın olay adı (örn. DISABLED_UPDATE)
 console.warn(`Hesap uyarısı: ${event.data.field} · ${event.data.event}`);
 break;
 default:
 console.log('Bilinmeyen event:', event.event);
 }

 processedEvents.add(eventId);
 res.json({ ok: true });
 } catch (err) {
 console.error('[WaWebhook] handler error:', err);
 res.status(500).json({ ok: false, error: 'handler_error' });
 }
});

// ═══════════════════════════════════════════════════════════════
// HANDLER'LAR — Kendi iş mantığınızı buraya yazın
// ═══════════════════════════════════════════════════════════════

function handleIncomingMessage(data) {
 // data.text düz metindir (medyada açıklama; açıklama yoksa boş).
 // Telefonu gizli kullanıcılarda data.from null gelir; kimlik data.user_id (BSUID) olur.
 const { from, user_id, name, type, text, media, phone_number_id } = data;
 const kimden = from || user_id || 'bilinmiyor';
 const ad = name || 'Müşteri';

 if (media && media.media_id) {
 console.log(`${ad} (${kimden}) ${type} gönderdi · media_id=${media.media_id}${text ? ' · ' + text : ''}`);
 // Dosyayı indirmek için (bkz. docs/10-media.md):
 // const { VeriMerkeziClient } = require('@verimerkezi/whatsapp-sdk');
 // const vm = new VeriMerkeziClient(process.env.VM_API_KEY);
 // vm.downloadMedia(media.media_id, require('fs').createWriteStream(`medya-${media.media_id}`)).catch(console.error);
 } else if (media && media.error) {
 // Nadiren dosya Meta'dan alınamaz; bu durumda indirilebilir dosya yoktur.
 console.warn(`${ad} (${kimden}) ${type} gönderdi, dosya alınamadı: ${media.error}`);
 } else {
 console.log(`${ad} (${kimden}): ${text}`);
 }

 // Otomatik yanıt göndermek için:
 // const { VeriMerkeziClient } = require('@verimerkezi/whatsapp-sdk');
 // const vm = new VeriMerkeziClient(process.env.VM_API_KEY);
 // await vm.sendText(phone_number_id, from, 'Mesajınızı aldık!');
}

function handleStatusUpdate(data, eventName) {
 const status = eventName.replace('message.status.', '');
 console.log(` ${data.wamid} -> ${status}`);
}

function handleFailedMessage(data) {
 // errors: Meta'nın hata listesi — [{ code, title, message, error_data: { details }, href }]
 const hata = (data.errors && data.errors[0]) || {};
 console.error(`Gönderim başarısız · ${data.wamid} · alıcı=${data.recipient} · kod=${hata.code ?? '-'} · ${hata.message || hata.title || 'bilinmiyor'}`);
}

app.get('/health', (req, res) => res.json({ status: 'ok' }));

app.listen(PORT, () => {
 console.log(`Veri Merkezi webhook receiver çalışıyor: http://localhost:${PORT}/wa-webhook`);
});
